src/disk_cache.rs

```diff
@@ -0,0 +1,956 @@
+        opts: PutOptions,
+    ) -> object_store::Result<PutResult> {
+        self.object_store.put_opts(location, payload, opts).await
+    }
+
+    async fn put_multipart(
+        &self,
+        location: &Path,
+    ) -> object_store::Result<Box<dyn MultipartUpload>> {
+        self.object_store.put_multipart(location).await
+    }
+
+    async fn put_multipart_opts(
+        &self,
+        location: &Path,
+        opts: PutMultipartOpts,
+    ) -> object_store::Result<Box<dyn MultipartUpload>> {
+        self.object_store.put_multipart_opts(location, opts).await
+    }
+
+    async fn delete(&self, location: &Path) -> object_store::Result<()> {
+        self.object_store.delete(location).await
+    }
+
+    fn list(&self, prefix: Option<&Path>) -> BoxStream<'_, object_store::Result<ObjectMeta>> {
+        self.object_store.list(prefix)
+    }
+
+    fn list_with_offset(
+        &self,
+        prefix: Option<&Path>,
+        offset: &Path,
+    ) -> BoxStream<'_, object_store::Result<ObjectMeta>> {
+        self.object_store.list_with_offset(prefix, offset)
+    }
+
+    async fn list_with_delimiter(&self, prefix: Option<&Path>) -> object_store::Result<ListResult> {
+        self.object_store.list_with_delimiter(prefix).await
+    }
+
+    async fn copy(&self, from: &Path, to: &Path) -> object_store::Result<()> {
+        self.object_store.copy(from, to).await
+    }
+
+    async fn rename(&self, from: &Path, to: &Path) -> object_store::Result<()> {
+        self.object_store.rename(from, to).await
+    }
+
+    async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> object_store::Result<()> {
+        self.object_store.copy_if_not_exists(from, to).await
+    }
+
+    async fn rename_if_not_exists(&self, from: &Path, to: &Path) -> object_store::Result<()> {
+        self.object_store.rename_if_not_exists(from, to).await
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct DiskCacheMeta {
+    pub location: String,
+    pub last_modified: String,
+    pub size: usize,
+    pub e_tag: Option<String>,
+    pub version: Option<String>,
+}
+
+// it seems that object_store did not expose this error type, duplicate it here.
+// TODO: raise a pr to expose this error type in object_store.
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum InvalidGetRange {
+    #[error("Range start too large, requested: {requested}, length: {length}")]
+    StartTooLarge { requested: usize, length: usize },
+
+    #[error("Range started at {start} and ended at {end}")]
+    Inconsistent { start: usize, end: usize },
+}
+
+struct DiskCacheEntry {
+    root_folder: std::path::PathBuf,
+    location: object_store::path::Path,
+    part_size: usize,
+}
+
+#[allow(unused)]
+type PartID = usize;
+
+#[allow(unused)]
+impl DiskCacheEntry {
+    /// save the GetResult to the disk cache, a GetResult may be transformed into multiple parts
+    /// file and a meta file. please note that the `range` in the GetResult is expected to be
```

Comment:
Nit: "part files"

And only "start" is expected to be ALIGNED, right?

Reply:
the end is also expected to be aligned, except the last part which might be smaller than the part size.

assume we've got an object file which is 3kb and part size is 1kb, we can pass `(2048, 2048 + 2)` to this function, it will mistakely cached a 2 bytes file as the 3rd part.

i think i can add an additional assert, with `object_size - end` must be aligned with `part_size`, as we can get the object_size in the `GetResult` meta.

Reply:
w.r.t the last file, sorry, yes that's what I meant. We don't do any assertions so any file could be misaligned. The logic you describe around part_size or object_size sounds right.